include/swift/AST/ReferenceStorage.def

```diff
@@ -0,0 +1,136 @@
+//===--- ReferenceStorage.def - Non-default reference storage ---*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines non-default reference storage kind macros used for
+// macro-metaprogramming.
+//
+//===----------------------------------------------------------------------===//
+
+/// There are two fundamental reference storage types: checked and unchecked.
+/// Checked storage types have runtime enforced correctness.
+/// Unchecked storage types have no runtime enforced correctness.
+///
+/// Checked reference storage types are also subcategorized by loadability.
+/// * Always loadable: The compiler may move the reference or use registers.
+/// * Never loadable: The runtime (etc) tracks the address of the reference.
+/// * Sometimes loadable: If the reference is a native object, then it is
+///   always loadable. Otherwise fall back to never loadable semantics.
+///
+/// Unchecked reference storage types are always loadable.
+///
+/// The primary macros therefore are:
+/// * ALWAYS_LOADABLE_CHECKED_REF_STORAGE
+/// * SOMETIMES_LOADABLE_CHECKED_REF_STORAGE
+/// * NEVER_LOADABLE_CHECKED_REF_STORAGE
+/// * UNCHECKED_REF_STORAGE
+///
+/// Helper macros include:
+/// * CHECKED_REF_STORAGE -- Any checked reference storage type.
+/// * LOADABLE_REF_STORAGE -- Any loadable reference storage type.
+/// * ALWAYS_OR_SOMETIMES_LOADABLE_CHECKED_REF_STORAGE
+/// * NEVER_OR_SOMETIMES_LOADABLE_CHECKED_REF_STORAGE
```

Comment:
Okay.  One of the things you should definitely document here is what each of these dimensions means in terms of various subsystems.  For example, "checked" storages have SIL instructions called `Load##Name##Inst` and `Store##Name##Inst`, always-or-sometimes-loadable storages have SIL instructions called `StrongRetain##Name##Inst`, and so on.

You should at least make sure it's possible to easily select all the storages in a category like that.

Reply:
I've added additional subsystem documentation in a forthcoming pull request update. Also, I thought this patch did "make sure it's possible to easily select all the storages in a category like that." What do you mean?